### ISSUE TITLE
Custom modifiers sometimes do not seem to be recognized

### ISSUE THREAD

**Issue by agocke** [dotnet/coreclr#10944]
When running Roslyn tests on CoreCLR I've hit a number of test failures around custom modifiers. I've examined the IL being generated by Roslyn as best I can and it seems to be valid. However, when CoreCLR attempts to run the binaries  it results in MissingMethodExceptions when attempting to call the methods which have custom modifiers.

I've attached a complete repro, including the CoreCLR 1.1 binaries and CoreFX dependencies, [here](https://microsoft-my.sharepoint.com/personal/angocke_microsoft_com/_layouts/15/guestaccess.aspx?guestaccesstoken=8%2bNOTWEv7ZGJ6S3dsEtVGXLUdhru2aJc6YQR9Aszpko%3d&docid=2_091bb9eca125e48749875f897abeb4d0e&rev=1).

The relevant code is simple. The first assembly has the following IL:

```
.class public auto ansi beforefieldinit Test1
       extends[mscorlib] System.Object
        {
  .method public hidebysig specialname rtspecialname
          instance void  .ctor() cil managed
        {
    // Code size       8 (0x8)
    .maxstack  8
    IL_0000:  ldarg.0
    IL_0001:  call instance void[mscorlib]
        System.Object::.ctor()
    IL_0006:  nop
    IL_0007:  ret
    } // end of method Test1::.ctor

  .method public hidebysig static void Test(valuetype[mscorlib] System.Nullable`1<int32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong)> x) cil managed
    {
    // Code size       11 (0xb)
    .maxstack  1
    IL_0000:  ldstr      ""Test""
    IL_0005:  call       void [mscorlib]System.Console::WriteLine(string)
    IL_000a:  ret
    } // end of method Test1::Test

} // end of class Test1
```

and the calling code in C# is:

```
class Module1
{
     static void Main()
    {
        Test1.Test(null);
    }
```

You can run the repro by cd'ing into the directory and calling `CoreRun.exe test2.dll`. `test2.dll` contains the C# code, while `_0_0d0741aa-1257-4a0f-850c-15c27e40a9dd.dll` contains the ilasm'd CIL.

**Comment by agocke** [comment]
cc @gkhanna79 @jkotas

**Comment by MichalStrehovsky** [comment]
```
`valuetype[mscorlib] System.Nullable`1<int32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong)>`
```

This is not permissible by the ECMA-335 standard. It seems like it's possible to express this in the format, but the behaviors would be undefined. Section II.23.2 of the spec defines when custom modifiers can show up within signatures: they can show up *at the beginning of* many of the signatures (and prefix a Type), but once you're in the middle of a Type signature, custom modifiers can be only applied to types pointed to by unmanaged pointers, or element types of SzArrays. In particular, generic arguments cannot be modified per the spec.

<img width="385" alt="param" src="https://cloud.githubusercontent.com/assets/13110571/25069154/eb8c38ee-222d-11e7-8fa4-2acb75170f0a.PNG">

<img width="427" alt="type" src="https://cloud.githubusercontent.com/assets/13110571/25069161/29f8e3f2-222e-11e7-88d4-9449615a5b60.PNG">

**Comment by agocke** [comment]
@MichalStrehovsky Yup, by a strict reading of the spec this is illegal. However, this feature is required to support seemingly simple features of C++ CLI like the following:

```csharp
    void Foo(Nullable<int> value)
    {
    }

    void Foo(Nullable<long> value)
    {
    }
```

There is a fairly large body of C++/CLI code in the wild that takes advantage of this feature. Roslyn encountered this same problem not too long ago and ended up altering our symbol model just to support this. See https://github.com/dotnet/roslyn/issues/4163 for the background on this issue. At the time we agreed that the CLI spec would be amended to support this feature, but I'm not sure what ever happened to the CLR side of this.

**Comment by yizhang82** [comment]
I don't think we support C++/CLI on CoreCLR.

**Comment by gkhanna79** [comment]
Correct - C++/CLI is not supported on .NET Core.

**Comment by kouvel** [comment]
Based on the above, it looks like there isn't anything to be done. Closing, please reactivate if I missed something.

**Comment by jkotas** [comment]
Roslyn tests are failing on CoreCLR because of this - https://github.com/dotnet/roslyn/issues/18411 . Don't we want to enable them?

Also, I do not think there was any intentional change between .NET Framework and .NET Core in this area. If the tests are passing on .NET Framework and failing on .NET Core, we should understand what changes at least.

**Comment by kouvel** [comment]
I see, ok
